contracts/State.sol

```diff
@@ -7,212 +7,231 @@ import './lib/verifier.sol';
    * @param state IDState commited
    */
-  event StateUpdated(uint256 id, uint64 blockN, uint64 timestamp, uint256 state);
-
-
-  function initState(
-            uint256 newState,
-            uint256 genesisState,
-            uint256 id,
-            uint[2] memory a,
-            uint[2][2] memory b,
-            uint[2] memory c
-  ) public {
-    require(identities[id].length==0);
-
-    _setState(newState, genesisState, id, a, b, c);
-  }
-
-  function setState(
-            uint256 newState,
-            uint256 id,
-            uint[2] memory a,
-            uint[2][2] memory b,
-            uint[2] memory c
-  ) public {
-    require(identities[id].length>0);
-
-    IDState memory oldIDState = identities[id][identities[id].length-1];
-    require(oldIDState.BlockN != block.number, "no multiple set in the same block");
-
-    _setState(newState, oldIDState.State, id, a, b, c);
-  }
-
-  function _setState(
-            uint256 newState,
-            uint256 oldState,
-            uint256 id,
-            uint[2] memory a,
-            uint[2][2] memory b,
-            uint[2] memory c
-  ) private {
-
-    uint256[3] memory input = [
-      id,
-      oldState,
-      newState
-    ];
-    require(verifier.verifyProof(a, b, c, input), "zkProof idState update could not be verified");
-    
-    identities[id].push(IDState(uint64(block.number), uint64(block.timestamp), newState));
-    emit StateUpdated(id, uint64(block.number), uint64(block.timestamp), newState);
-  }
-
-  /**
-   * Retrieve last state for a given identity
-   * @param id identity
+    event StateUpdated(uint256 id, uint64 blockN, uint64 timestamp, uint256 state);
+
+
+    function initState(
+        uint256 newState,
+        uint256 genesisState,
+        uint256 id,
+        uint[2] memory a,
+        uint[2][2] memory b,
+        uint[2] memory c
+    ) public {
+        require(identities[id].length==0);
+
+        _setState(newState, genesisState, id, a, b, c);
+    }
+
+
+    function setState(
+        uint256 newState,
+        uint256 id,
+        uint[2] memory a,
+        uint[2][2] memory b,
+        uint[2] memory c
+    ) public {
+        require(identities[id].length>0);
+
+        IDState memory oldIDState = identities[id][identities[id].length-1];
+        require(oldIDState.BlockN != block.number, "no multiple set in the same block");
+
+        _setState(newState, oldIDState.State, id, a, b, c);
+    }
+
+    function _setState(
+        uint256 newState,
+        uint256 oldState,
+        uint256 id,
+        uint[2] memory a,
+        uint[2][2] memory b,
+        uint[2] memory c
+    ) private {
+
+        uint256[3] memory input = [
+        id,
+        oldState,
+        newState
+        ];
+        require(verifier.verifyProof(a, b, c, input), "zkProof idState update could not be verified");
+
+        identities[id].push(IDState(uint64(block.number), uint64(block.timestamp), newState));
+
+        transitions[oldState] = now;
```

Comment:
Probably there should be `newState`, not `oldState`, or variable name should be renamed.

Reply:
maybe `phasedOutAt` or `replacedAt`

Reply:
transitions map stores timestamp when old state of identity transited to new state.

Reply:
maybe name can be changed in next iteration but it will require integration changes

Reply:
this mapping will be changed in the nearest feature, so we can take an appropriate name.